File: C++/GPT-4/Queue/Easy/Solution/2073. Time Needed to Buy Tickets/2.cpp
#include <vector>
#include <queue>
#include <utility>

using namespace std;

class Solution {
public:
    int timeRequiredToBuy(vector<int>& tickets, int k) {
        int n = tickets.size(), time = 0;
        queue<pair<int, int>> q; // store the original order
        priority_queue<pair<int, int>> pq; // store the tickets and index

        // Initialize the queues
        for(int i = 0; i < n; ++i) {
            q.push({tickets[i], i});
            pq.push({tickets[i], i});
        }

        while(true){
            auto top = pq.top();
            // Sell tickets in the original order
            while(top.first != q.front().first || top.second != q.front().second){
                auto cur = q.front();
                q.pop();
                q.push({cur.first, cur.second});
            }
            
            // Sell one ticket
            q.pop();
            pq.pop();
            time++;

            // If it's the kth person and no tickets left, break
            if(top.second == k && top.first == 1)
                break;
            
            // If there're still tickets left, push the person back to the queue
            if(top.first > 1)
                pq.push({top.first - 1, top.second});
        }

        return time;
    }
};

File: Java/GPT-4/Queue/Easy/Solution/2073. Time Needed to Buy Tickets/2.java
import java.util.*;

class Solution {
    public int timeRequiredToBuy(int[] tickets, int k) {
        int n = tickets.length;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            queue.offer(new int[] {tickets[i], i});
        }
        int time = 0;
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            time++;
            if (current[0] > 1) {
                current[0]--;
                queue.offer(current);
            }
            if (current[1] == k && current[0] == 1) {
                break;
            }
        }
        return time;
    }
}

